src/accumulator/proof.rs

```diff
@@ -0,0 +1,310 @@
+    ///```
+    pub fn verify(&self, del_hashes: &Vec<sha256::Hash>, stump: &Stump) -> Result<bool, String> {
+        let calculated_roots = self.create_root_candidates(del_hashes, stump)?;
+        let stump_roots = util::num_roots(stump.leafs);
+        let mut number_matched_roots = 0;
+
+        for i in 0..stump_roots {
+            if calculated_roots[number_matched_roots] == stump.roots[stump.roots.len() - (i + 1)] {
+                number_matched_roots += 1;
+                if number_matched_roots == calculated_roots.len() {
+                    break;
+                }
+            }
+        }
+
+        if calculated_roots.len() != number_matched_roots && calculated_roots.len() != 0 {
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    pub fn create_root_candidates(
+        &self,
+        del_hashes: &Vec<sha256::Hash>,
+        stump: &Stump,
+    ) -> Result<Vec<sha256::Hash>, String> {
+        // Where all the root hashes that we've calculated will go to.
+        let total_rows = util::tree_rows(stump.leafs);
+        let mut targets = self.targets.clone();
+
+        // Where all the parent hashes we've calculated in a given row will go to.
+        let mut calculated_root_hashes =
+            Vec::<sha256::Hash>::with_capacity(util::num_roots(stump.leafs) as usize);
+
+        // As we calculate nodes upwards, it accumulates here
+        let mut nodes: Vec<_> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(del_hashes.to_owned())
+            .collect();
+
+        let mut row = 0;
+        let mut proof_hashes_index = 0 as usize;
+
+        while row <= total_rows {
+            let res = util::detect_row_hashes(&targets, row, total_rows);
+
+            if let Ok((start, end)) = res {
+                let mut i = start as usize;
+
+                // Iterates over a row
+                while i <= end as usize {
+                    if nodes.len() == 0 {
+                        return Err(String::from("Malformed tree"));
+                    }
+
+                    // If the current position is a root, we add that to our result and don't go any further
+                    if util::is_root_position(nodes[i].0, stump.leafs, total_rows) {
+                        calculated_root_hashes.push(nodes[i].1);
+                    } else if nodes.len() > i + 1 && (nodes[i].0 | 1 == nodes[i + 1].0) {
+                        // If the next node is the current node's sibling, we hash both nodes together
+
+                        let hash = types::parent_hash(&nodes[i].1, &nodes[i + 1].1);
+                        let next_to_prove = util::parent(nodes[i].0, total_rows);
+
+                        targets.push(next_to_prove);
+                        targets.sort();
+
+                        Proof::sorted_push(&mut nodes, (next_to_prove, hash));
+
+                        i += 1;
```

Comment:
Could we comment here that we're doing an additional `i += 1` because we've processed an extra node in this loop?